pkg/registry/generic/registry/store.go

```diff
@@ -513,34 +519,54 @@ func shouldUpdateFinalizers(e *Store, accessor meta.Object, options *api.DeleteO
 	}
 	// If a finalizer is set in the object, it overrides the default
 	hasOrphanFinalizer := false
+	hasSynchronousGCFinalizer := false
 	finalizers := accessor.GetFinalizers()
 	for _, f := range finalizers {
 		if f == api.FinalizerOrphan {
 			shouldOrphan = true
 			hasOrphanFinalizer = true
-			break
+		}
+		if f == api.FinalizerSynchronousGC {
+			shouldSynchronousGC = true
+			hasSynchronousGCFinalizer = true
 		}
 		// TODO: update this when we add a finalizer indicating a preference for the other behavior
 	}
 	// If an explicit policy was set at deletion time, that overrides both
 	if options != nil && options.OrphanDependents != nil {
 		shouldOrphan = *options.OrphanDependents
 	}
+	if options != nil && options.SynchronousGarbageCollection != nil {
+		shouldSynchronousGC = *options.SynchronousGarbageCollection
+		if *options.SynchronousGarbageCollection && options.OrphanDependents == nil {
+			//default to not orphan if the option is nil, regardless of the
+			//existing finalizers or the DefaultGarbageCollectionPolicy().
+			shouldOrphan = false
+		}
+	}
+	// compute shouldUpdate and newFinalizers
```

Comment:
Can I get a more precise comment?  Something like  `// if we need a finalizer and that finalizer doesn't already exist on the object, we should update it`.